### PR TITLE
[Keyboard] Add Olly_Octagon PCB to QMK

### PR DESCRIPTION
**Pull request by mechlovin**
## Description

Add support for Duck Octagon V2 and V3 custom keyboards.
Thanks.
## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [ ] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/mechlovin/olly/octagon/config.h [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+#pragma once
+
+#include "config_common.h"
+
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID    0x4D4C
+#define PRODUCT_ID   0xD750 
+#define DEVICE_VER   0x0002
+#define PRODUCT      Olly Octagon
```
Comment (waffle87) [inline keyboards/mechlovin/olly/octagon/config.h:27]: This needs to be moved to `info.json`.

File: keyboards/mechlovin/olly/octagon/config.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+/* COL2ROW, ROW2COL */
+#define DIODE_DIRECTION COL2ROW
+
+/* key matrix size */
+#define MATRIX_ROWS 6
+#define MATRIX_COLS 16
+
+#define MATRIX_ROW_PINS { B0, B12, A6, A5, A4, A3 }
+#define MATRIX_COL_PINS { A10, A9, A8, B15, B14, B13, B2, B1, A15, B3, B9, B8, B7, B6, B5, B4 }
+#define UNUSED_PINS
```
Comment (waffle87) [inline keyboards/mechlovin/olly/octagon/config.h:38]:
```suggestion
```

File: keyboards/mechlovin/olly/octagon/config.h [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+#define ENABLE_LED_MATRIX_SOLID_REACTIVE_MULTICROSS	
+#define ENABLE_LED_MATRIX_SOLID_REACTIVE_NEXUS	
+#define ENABLE_LED_MATRIX_SOLID_REACTIVE_MULTINEXUS	
+#define ENABLE_LED_MATRIX_SOLID_SPLASH	
+#define ENABLE_LED_MATRIX_SOLID_MULTISPLASH	
+
+#define I2C1_CLOCK_SPEED 400000
+#define I2C1_DUTY_CYCLE FAST_DUTY_CYCLE_2
+
+#define VIA_QMK_LED_MATRIX_ENABLE
```
Comment (waffle87) [inline keyboards/mechlovin/olly/octagon/config.h:99]:
```suggestion
```

File: keyboards/mechlovin/olly/octagon/octagon.c [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+       1,            1,             1,             1,             1,             1,             1,             1,             1,             1,             1,             1,             1,             1,                            1,
+       1,            1,             1,             1,             1,             1,             1,             1,             1,             1,             1,             1,             1,             1,                            1,
+       1,            1,             1,             1,             1,             1,             1,             1,             1,             1,             1,             1,             1,             1,                            1,
+       1,            1,             1,                                                          1,                                                          1,             1,             1,             1,             1,             1,
+    } 
+};
+
+
+__attribute__((weak)) void led_matrix_indicators_kb(void) {
```
Comment (waffle87) [inline keyboards/mechlovin/olly/octagon/octagon.c:62]:
```suggestion
bool led_matrix_indicators_kb(void) {
    if (!led_matrix_indicators_user()) { return false; }
```

File: keyboards/mechlovin/olly/octagon/octagon.c [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    if (host_keyboard_led_state().num_lock) {
+        led_matrix_set_value(88, 0xFF);
+    } else {
+        led_matrix_set_value(88, 0x00);
+    }
+    if (host_keyboard_led_state().scroll_lock) {
+        led_matrix_set_value(89, 0xFF);
+    } else {
+        led_matrix_set_value(89, 0x00);
+    }
```
Comment (waffle87) [inline keyboards/mechlovin/olly/octagon/octagon.c:78]:
```suggestion
    }
    return true;
```

File: keyboards/mechlovin/olly/octagon/octagon.c [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+
+__attribute__((weak))
+layer_state_t layer_state_set_user(layer_state_t state) {
```
Comment (waffle87) [inline keyboards/mechlovin/olly/octagon/octagon.c:83]:
```suggestion
layer_state_t layer_state_set_kb(layer_state_t state) {
```

File: keyboards/mechlovin/olly/octagon/octagon.c [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+  // if on layer 4, turn on L4 LED, otherwise off.
+    if (get_highest_layer(state) == 3) {
+        led_matrix_set_value(93, 0xFF);
+    } else {
+        led_matrix_set_value(93, 0x00);
+    }
+
+    return state;
```
Comment (waffle87) [inline keyboards/mechlovin/olly/octagon/octagon.c:111]:
```suggestion
    return layer_state_set_user(state);
```

**Review by waffle87 — CHANGES_REQUESTED** [review]

File: keyboards/mechlovin/olly/octagon/config.h [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+// 0b1110100 AD <-> GND
+// 0b1110111 AD <-> VCC
+// 0b1110101 AD <-> SCL
+// 0b1110110 AD <-> SDA
+#define LED_DRIVER_ADDR_1 0b1110110
+#define I2C_DRIVER I2CD2
+#define I2C1_SCL_PIN B10
+#define I2C1_SDA_PIN B11
+#define LED_DRIVER_COUNT 1
+#define DRIVER_LED_TOTAL 94
```
Comment (drashna) [inline keyboards/mechlovin/olly/octagon/config.h:70]:
```suggestion
#define LED_MATRIX_LED_COUNT 94
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

**Review by keyboard-magpie — APPROVED** [review]

**Pull request merged**